optuna/integration/pytorch_lightning.py

```diff
@@ -69,60 +72,86 @@ def on_fit_start(self, trainer: Trainer, pl_module: "pl.LightningModule") -> Non
+                f"The metric '{self.monitor}' is not in the evaluation logs for pruning. "
+                "Please make sure you set the correct metric name."
             )
             warnings.warn(message)
             return
 
+        epoch = pl_module.current_epoch
         should_stop = False
+
+        # Determine if the trial should be terminated in a single process.
+        if not self.is_ddp_backend:
+            self._trial.report(current_score.item(), step=epoch)
+            should_stop = self._trial.should_prune()
+            if not should_stop:
+                return
+            message = f"Trial was pruned at epoch {epoch}."
+            raise optuna.TrialPruned(message)
+
+        # Determine if the trial should be terminated in a DDP.
         if trainer.is_global_zero:
             self._trial.report(current_score.item(), step=epoch)
             should_stop = self._trial.should_prune()
+
+            # Update intermediate value in the storage.
+            _trial_id = self._trial._trial_id
+            _study = self._trial.study
+            _trial_system_attrs = _study._storage.get_trial_system_attrs(_trial_id)
+            intermediate_values = _trial_system_attrs.get(_INTERMEDIATE_VALUE)
+            intermediate_values[epoch] = current_score.item()  # type: ignore[index]
+            self._trial.storage.set_trial_system_attr(
+                self._trial._trial_id, _INTERMEDIATE_VALUE, intermediate_values
+            )
+
+        # Terminate every process if any world process decides to stop.
         should_stop = trainer.strategy.broadcast(should_stop)
+        trainer.should_stop = trainer.should_stop or should_stop
         if not should_stop:
             return
 
-        if not self.is_ddp_backend:
-            message = "Trial was pruned at epoch {}.".format(epoch)
-            raise optuna.TrialPruned(message)
-        else:
-            # Stop every DDP process if global rank 0 process decides to stop.
-            trainer.should_stop = True
-            if trainer.is_global_zero:
-                self._trial.storage.set_trial_system_attr(self._trial._trial_id, _PRUNED_KEY, True)
-                self._trial.storage.set_trial_system_attr(self._trial._trial_id, _EPOCH_KEY, epoch)
-
-    def on_fit_end(self, trainer: Trainer, pl_module: LightningModule) -> None:
-        if not self.is_ddp_backend:
-            return
+        if trainer.is_global_zero:
+            # Update system_attr from global zero process.
+            self._trial.storage.set_trial_system_attr(self._trial._trial_id, _PRUNED_KEY, True)
+            self._trial.storage.set_trial_system_attr(self._trial._trial_id, _EPOCH_KEY, epoch)
+
+    def check_pruned(self) -> None:
+        """Raise :class:`optuna.TrialPruned` manually if pruned.
+
+        Currently, intermediate_values are not properly propagated between processes due to
+        _CachedStorage. Therefore, necessary information is kept in trial_system_attrs when the
```

Comment:
`_CachedStorage` is a private class, so how about mentioning the storage cache without showing the class name?

```suggestion
       storage cache. Therefore, necessary information is kept in trial_system_attrs when the
```